pkg/lib/controller-runtime/source/dynamic.go

```diff
@@ -0,0 +1,196 @@
+func (d *Dynamic) Start(handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) error {
+	if d.client == nil {
+		return fmt.Errorf("must inject rest config before starting dynamic source")
+	}
+	if d.cache == nil {
+		return fmt.Errorf("must inject cache before starting dynamic source")
+	}
+	if d.stop == nil {
+		return fmt.Errorf("must inject stop channel before starting dynamic source")
+	}
+
+	// TODO(njhale): Stop sources when GVKs are removed
+	var (
+		synced sync.WaitGroup
+		once   sync.Once
+	)
+	synced.Add(1)
+
+	go wait.Until(func() {
+		defer once.Do(func() {
+			synced.Done()
+		})
+
+		informable, err := d.InformableGVKs()
+		if err != nil {
+			log.Error(err, "could not get available gvks")
+		}
+
+		for _, gvk := range informable {
+			if d.sourceStarted(gvk) {
+				log.V(4).Info("dynamic source already started", "gvk", gvk)
+				continue
+			}
+
+			if err := d.startSource(gvk, handle, queue, predicates...); err != nil {
+				log.Error(err, "failed to start dynamic source", "gvk", gvk)
+			}
+		}
+
+	}, 15*time.Second, d.stop)
+
+	synced.Wait()
+
+	return nil
+}
+
+var (
+	knownInformables []schema.GroupVersionKind
+	informablesLock  sync.RWMutex
+)
+
+// InformableGVKs returns the set of GVKs available in discovery that support the list and watch verbs (informable).
+func (d *Dynamic) InformableGVKs() (informables []schema.GroupVersionKind, err error) {
+	if d.client == nil {
+		return nil, fmt.Errorf("no client")
+	}
+
+	var resourceLists []*metav1.APIResourceList
+	resourceLists, err = d.client.ServerPreferredResources()
+	if err != nil {
+		informablesLock.RLock()
+		defer informablesLock.RUnlock()
+		log.Error(err, "could not get preferred resources, returning latest known gvks")
+		return knownInformables, nil
+	}
+
+	latest := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"list", "watch"}}, resourceLists)
+	for _, rl := range latest {
+		gv, err := schema.ParseGroupVersion(rl.GroupVersion)
+		if err != nil {
+			log.Error(err, "could not parse group version")
+			continue
+		}
+
+		for _, r := range rl.APIResources {
+			gvk := schema.GroupVersionKind{Group: gv.Group, Version: gv.Version, Kind: r.Kind}
+			informables = append(informables, gvk)
+		}
+	}
+
+	if len(informables) > 0 {
+		// Memoize the latest informables in case discovery fails on successive calls
+		informablesLock.Lock()
+		defer informablesLock.Unlock()
+		knownInformables = informables
+	}
+
+	return
+}
+
+func (d *Dynamic) startSource(gvk schema.GroupVersionKind, handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) (err error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.sources == nil {
+		d.sources = map[schema.GroupVersionKind]source.Source{}
+	}
+
+	r := &unstructured.Unstructured{}
+	r.SetGroupVersionKind(gvk)
+	s := &source.Kind{Type: r}
+	if err = s.InjectCache(d.cache); err != nil {
```

Comment:
controller-runtime doesn't expose a way to stop individual sources/informers, see:

- https://github.com/kubernetes-sigs/controller-runtime/issues/935
- https://github.com/kubernetes-sigs/controller-runtime/pull/936
- https://github.com/gardener/gardener-resource-manager/issues/61

I'm curious to see if anyone has suggestions for how to get around this short of reimplementing controller-runtime's [informerCache](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/informer_cache.go).